File: reference/windows-winui/AirX/Services/AirXCloud.cs
using AnyDrop.Model;
using AnyDrop.Util;
using Newtonsoft.Json;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Nodes;
using System.Threading.Tasks;
using Windows.ApplicationModel.Appointments;
using Windows.Data.Json;

namespace AnyDrop.Services
{
    public static class AnyDropCloud
    {
#if DEBUG
        public const string ApiBaseUrl = "https://anydrop.eggtartc.com";
        public const string WebSocketBaseUrl = "wss://anydrop.eggtartc.com/device-register";
        // public const string ApiBaseUrl = "http://10.0.0.205:2479";
        // public const string WebSocketBaseUrl = "ws://10.0.0.205:2479/device-register";
#else
        public const string ApiBaseUrl = "https://anydrop.eggtartc.com";
        public const string WebSocketBaseUrl = "wss://anydrop.eggtartc.com/device-register";
#endif

        public class UnauthorizedException : Exception { }
        public class IncorrectCredentialTypeException : Exception { }

        private static string GetParametersFromDictionary(Dictionary<string, dynamic> parameters)
        {
            StringBuilder builder = new();
            foreach (var entry in parameters)
            {
                builder.Append(entry.Key);
                builder.Append("=");
                builder.Append(entry.Value);
                builder.Append("&");
            }
            builder.Length--;
            return builder.ToString();
        }

        private static async Task<T> RequestAsync<T>(
            HttpMethod method,
            string path,
            Dictionary<string, dynamic> body,
            bool needToken
        )
        {
            var client = new HttpClient();

            HttpRequestMessage request = null;
            if (method == HttpMethod.Post)
            {
                request = new HttpRequestMessage(method, ApiBaseUrl + path);
                var jsonBody = JsonConvert.SerializeObject(body);
                var content = new StringContent(jsonBody, null, "application/json");
                request.Content = content;
            }
            else
            {
                string fullPath = ApiBaseUrl + path;
                if (body != null)
                {
                    fullPath += "?" + GetParametersFromDictionary(body);
                }
                request = new HttpRequestMessage(method, fullPath);
            }

            if (needToken)
            {
                if (SettingsUtils.ReadCredentialType() != CredentialType.AnyDropToken)
                {
                    throw new IncorrectCredentialTypeException();
                }
                var credential = SettingsUtils.String(Keys.SavedCredential, "");
                request.Headers.Authorization = new("Bearer", credential);
            }

            var response = await client.SendAsync(request);
            switch (response.StatusCode)
            {
                case System.Net.HttpStatusCode.Unauthorized:
                    {
                        throw new UnauthorizedException();
                    }
                case System.Net.HttpStatusCode.OK:
                    {
                        string responseString = await response.Content.ReadAsStringAsync();
                        return JsonConvert.DeserializeObject<T>(responseString);
                    }
                default:
                    {
                        throw new HttpRequestException("Status code = " + response.StatusCode);
                    }
            }
        }

        public static async Task<LoginResponse> LoginAsync(string uid, string password, string salt = "114514")
        {
            password = HashUtils.Sha256(HashUtils.Sha256(password));
            return await RequestAsync<LoginResponse>(
                HttpMethod.Post,
                "/auth/token",
                new()
                {
                    { "uid", uid },
                    { "password", password },
                    { "salt", salt }
                },
                false
            );
        }

        public static async Task<RenewResponse> RenewAsync(string uid)
        {
            return await RequestAsync<RenewResponse>(
                HttpMethod.Post,
                "/auth/renew",
                new()
                {
                    { "uid", uid }
                },
                true
            );
        }

        public static async Task<GreetingsResponse> GreetingsAsync()
        {
            return await RequestAsync<GreetingsResponse>(
                HttpMethod.Get,
                "/api/v1/greetings",
                null,
                true
            );
        }

        public static async Task<MessageSendResponse> SendMessageAsync(string content, MessageType type)
        {
            return await RequestAsync<MessageSendResponse>(
                HttpMethod.Post,
                 "/api/v1/message",
                new()
                {
                    { "content", content },
                    { "type", (int)type }
                },
                 true
            );
        }

        public class LoginResponse
        {
            public bool success;
            public string message;
            public string name;
            public string token;
        }

        public class RenewResponse
        {

            public bool success;
            public string message;
            public string token;
        }

        public class GreetingsResponse
        {
            public bool success;
            public string message;
            public string name;
            public int uid;
        }

        public class MessageSendResponse
        {
            public bool success;
            public string message;
        }
    }
}

File: reference/macos-swiftui/AirX/Service/AirXCloud.swift
//
//  NetworkUtil.swift
//  AnyDropmac
//
//  Created by Hatsune Miku on 2023-06-14.
//

import Foundation
import Alamofire

class AnyDropCloud {
#if DEBUG
    // 没事儿暂时不动后端
    public static let API_BASE = "https://anydrop.eggtartc.com"
    public static let WEBSOCKET_BASE = "wss://anydrop.eggtartc.com/device-register"

    // public static let API_BASE = "http://localhost:2479"
    // public static let WEBSOCKET_BASE = "http://localhost:2479/device-register"
#else
    public static let API_BASE = "https://anydrop.eggtartc.com"
    public static let WEBSOCKET_BASE = "wss://anydrop.eggtartc.com/device-register"
#endif

    enum AnyDropError: Error {
        case hashAlgorithmUnavailable
        case incorrectCredential
    }
    
    struct LoginPacket: Encodable {
        let uid: String
        let password: String
        let salt: String
    }
    
    struct LoginResponse: Decodable {
        let success: Bool
        let message: String
        let name: String?
        let token: String?
    }
    
    struct RenewPacket: Encodable {
        let uid: String
    }
    
    struct RenewResponse: Decodable {
        let success: Bool
        let message: String
        let token: String?
    }
    
    struct MessageSendPacket: Encodable {
        let content: String
        let type: Int
    }
    
    struct MessageSendResponse: Decodable {
        let success: Bool
        let message: String
    }
    
    public static func login(
        uidOrEmail: String,
        password: String,
        completion: @escaping (_ response: LoginResponse) -> Void
    ) throws {
        let salt = "114514"
        guard let passwordSha256Sha256 = password.sha256()?.sha256() else {
            throw AnyDropError.hashAlgorithmUnavailable
        }
        let packet = LoginPacket(uid: uidOrEmail, password: passwordSha256Sha256, salt: salt)
        try post(
            "/auth/token",
            parameters: packet,
            requireAuthentication: false,
            completion: completion
        )
    }
    
    public static func renew(
        uid: String,
        completion: @escaping (_ response: RenewResponse) -> Void
    ) throws {
        let packet = RenewPacket(uid: uid)
        try post(
            "/auth/renew",
            parameters: packet,
            requireAuthentication: true,
            completion: completion
        )
    }
    
    public static func sendMessage(
        content: String,
        type: MessageType,
        completion: @escaping (_ response: MessageSendResponse) -> Void
    ) throws {
        let packet = MessageSendPacket(content: content, type: type.rawValue)
        try post(
            "/api/v1/message",
            parameters: packet,
            requireAuthentication: true,
            completion: completion
        )
    }
    
    private static func post<P: Encodable, T: Decodable>(
        _ path: String,
        parameters: P,
        requireAuthentication: Bool,
        // TODO: learn @escaping
        completion: @escaping (_ response: T) -> Void
    ) throws {
        var headers = HTTPHeaders.default
        if requireAuthentication {
            // Is anydrop token?
            if Defaults.credentialType() != .anydropToken {
                throw AnyDropError.incorrectCredential
            }

            // Is token seems valid?
            let token = Defaults.string(.savedCredential, def: "")
            guard !token.isEmpty else {
                throw AnyDropError.incorrectCredential
            }

            headers.add(.authorization("Bearer \(token)"))
        }
        
        AF.request(
            API_BASE + path,
            method: .post,
            parameters: parameters,
            encoder: JSONParameterEncoder.default,
            headers: headers
        ).responseDecodable(of: T.self) { decoded in
            if let decodedValue = decoded.value {
                DispatchQueue.main.async {
                    completion(decodedValue)
                }
            }
            else if let error = decoded.error {
                AccountUtils.clearSavedUserInfoAndSignOut()
                AccountUtils.notifySubscribers(didLoginSuccess: false)
                print("Error in decoding response: \(error), path: \(path)")
            }
            else {
                print("Error in decoding response: Unknown error")
            }
        }
    }
}
